pkg/agent/nodeportlocal/npl_agent_init.go

```diff
@@ -55,39 +53,32 @@ func InitializeNPLAgent(kubeClient clientset.Interface, portRange, nodeName stri
 	if err != nil {
 		return nil, fmt.Errorf("NPL rules for pod ports could not be initialized, error: %v", err)
 	}
-	return InitController(kubeClient, portTable, nodeName)
+	return InitController(kubeClient, informerFactory, portTable, nodeName)
 }
 
-// InitController creates an Informer which watches for events from Pods running on the same Node where Antea agent is running.
+// InitController initializes the NPLController with appropriate Pod and Service Informers.
 // This function can be used independently while unit testing without using InitializeNPLAgent function.
-func InitController(kubeClient clientset.Interface, portTable *portcache.PortTable, nodeName string) (*k8s.Controller, error) {
-	c := k8s.NewNPLController(kubeClient, portTable)
-	c.RemoveNPLAnnotationFromPods()
-
+func InitController(kubeClient clientset.Interface, informerFactory informers.SharedInformerFactory, portTable *portcache.PortTable, nodeName string) (*k8s.NPLController, error) {
 	// Watch only the Pods which belong to the Node where the agent is running
-	fieldSelector := fields.OneTermEqualSelector("spec.nodeName", nodeName)
-	optionsModifier := func(options *metav1.ListOptions) {
-		options.FieldSelector = fieldSelector.String()
+	listOptions := func(options *metav1.ListOptions) {
+		options.FieldSelector = fields.OneTermEqualSelector("spec.nodeName", nodeName).String()
 	}
-	lw := &cache.ListWatch{
-		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
-			optionsModifier(&options)
-			return kubeClient.CoreV1().Pods(metav1.NamespaceAll).List(context.TODO(), options)
-		},
-		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
-			optionsModifier(&options)
-			return kubeClient.CoreV1().Pods(metav1.NamespaceAll).Watch(context.TODO(), options)
-		},
-	}
-
-	cacheStore, controller := cache.NewInformer(lw, &corev1.Pod{}, resyncPeriod,
-		cache.ResourceEventHandlerFuncs{
-			AddFunc:    c.EnqueueObj,
-			DeleteFunc: c.EnqueueObj,
-			UpdateFunc: func(old, cur interface{}) { c.EnqueueObj(cur) },
-		},
+	podInformer := coreinformers.NewFilteredPodInformer(
+		kubeClient,
+		metav1.NamespaceAll,
+		resyncPeriod,
+		cache.Indexers{},
+		listOptions,
 	)
-	c.CacheStore = cacheStore
-	c.Ctrl = controller
+
+	svcInformer := informerFactory.Core().V1().Services().Informer()
+
+	c := k8s.NewNPLController(kubeClient,
+		podInformer,
+		svcInformer,
+		resyncPeriod,
+		portTable)
+	c.RemoveNPLAnnotationFromPods()
```

Comment:
Sorry, but why we call this func here?

Reply:
During restart we are removing all IPTABLES rules and annotations, so that we can populate all rules from informer  add/update events and there is no stale entry, in case a POD is deleted. 
This has been improved in the PR for restartability, where we selectively add/delete annotations, instead of deleting everything.

Reply:
And maybe add some comments to explain it and saying it is a workaround.

Reply:
This is supposed to go away as part of https://github.com/vmware-tanzu/antrea/pull/1758, which makes improvements in this logic.